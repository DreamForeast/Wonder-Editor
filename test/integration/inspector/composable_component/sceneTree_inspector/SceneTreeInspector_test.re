open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("SceneTreeInspector", () => {
    let sandbox = getSandboxDefaultVal();
    let _getFromArray = (array, index) => ArrayService.getNth(index, array);
    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initStateAndGl(~sandbox, ());
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));
    describe("prepare currentSelectSource is SceneTree", () => {
      beforeEach(() =>
        CurrentSelectSourceEditorService.setCurrentSelectSource(
          EditorType.SceneTree,
        )
        |> StateLogicService.getAndSetEditorState
      );

      describe("test show component", () => {
        test("if hasn't currentSceneTreeNode, show nothing", () => {
          MainEditorSceneTool.createDefaultScene(
            sandbox,
            MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
          );

          BuildComponentTool.buildInspectorComponent(
            TestTool.buildEmptyAppState(),
            InspectorTool.buildFakeAllShowComponentConfig(),
          )
          |> ReactTestTool.createSnapshotAndMatch;
        });

        describe("else", () => {
          test(
            "if currentSceneTreeNode is camera, should show transform and basicCameraView and perspectiveCameraProjection",
            () => {
              MainEditorSceneTool.createDefaultScene(
                sandbox,
                MainEditorSceneTool.setCameraTobeCurrentSceneTreeNode,
              );
              BuildComponentTool.buildInspectorComponent(
                TestTool.buildEmptyAppState(),
                InspectorTool.buildFakeAllShowComponentConfig(),
              )
              |> ReactTestTool.createSnapshotAndMatch;
            },
          );

          test(
            "else if currentSceneTreeNode is box, should show transform and basicMaterial",
            () => {
            MainEditorSceneTool.createDefaultScene(
              sandbox,
              MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
            );
            BuildComponentTool.buildInspectorComponent(
              TestTool.buildEmptyAppState(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
        });
        describe("test add component workflow", () => {
          beforeEach(() =>
            MainEditorSceneTool.createDefaultScene(
              sandbox,
              MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
            )
          );
          test(
            "click the add component button, show addableComponent list", () => {
            let component =
              BuildComponentTool.buildInspectorComponent(
                TestTool.buildEmptyAppState(),
                InspectorTool.buildFakeAllShowComponentConfig(),
              );
            BaseEventTool.triggerComponentEvent(
              component,
              OperateComponentEventTool.triggerClickAddComponentEvent,
            );
            component |> ReactTestTool.createSnapshotAndMatch;
          });
          test("click sourceInstance component, add to inspector", () => {
            let component =
              BuildComponentTool.buildInspectorComponent(
                TestTool.buildEmptyAppState(),
                InspectorTool.buildFakeAllShowComponentConfig(),
              );
            BaseEventTool.triggerComponentEvent(
              component,
              OperateComponentEventTool.triggerClickAddComponentEvent,
            );
            BaseEventTool.triggerComponentEvent(
              component,
              OperateComponentEventTool.triggerClickAddSourceInstanceEvent,
            );
            let component2 =
              BuildComponentTool.buildInspectorComponent(
                TestTool.buildEmptyAppState(),
                InspectorTool.buildFakeAllShowComponentConfig(),
              );
            component2 |> ReactTestTool.createSnapshotAndMatch;
          });
        });
      });
      describe("test gameObject rename", () => {
        beforeEach(() =>
          MainEditorSceneTool.createDefaultScene(
            sandbox,
            MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
          )
        );
        describe("test snapshot", () => {
          test("test rename to specific name", () => {
            let newName = "gameObjectNewName";
            let inspectorComponent =
              BuildComponentTool.buildInspectorComponent(
                TestTool.buildAppStateSceneGraphFromEngine(),
                InspectorTool.buildFakeAllShowComponentConfig(),
              );
            BaseEventTool.triggerComponentEvent(
              inspectorComponent,
              GameObjectRenameTool.triggerRenameChangeEvent(newName),
            );
            BaseEventTool.triggerComponentEvent(
              inspectorComponent,
              GameObjectRenameTool.triggerRenameBlurEvent(newName),
            );
            BuildComponentTool.buildInspectorComponent(
              TestTool.buildAppStateSceneGraphFromEngine(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
          describe("deal with specific case", () =>
            test(
              "ket in '', trigger onBlur, the input value should be original name",
              () => {
              let newName = "";
              let inspectorComponent =
                BuildComponentTool.buildInspectorComponent(
                  TestTool.buildAppStateSceneGraphFromEngine(),
                  InspectorTool.buildFakeAllShowComponentConfig(),
                );
              BaseEventTool.triggerComponentEvent(
                inspectorComponent,
                GameObjectRenameTool.triggerRenameChangeEvent(newName),
              );
              BaseEventTool.triggerComponentEvent(
                inspectorComponent,
                GameObjectRenameTool.triggerRenameBlurEvent(newName),
              );
              BuildComponentTool.buildInspectorComponent(
                TestTool.buildAppStateSceneGraphFromEngine(),
                InspectorTool.buildFakeAllShowComponentConfig(),
              )
              |> ReactTestTool.createSnapshotAndMatch;
            })
          );
        });
        describe("test logic", () =>
          describe("test engine", () =>
            test("test rename gameObject ", () => {
              let newName = "gameObjectNewName";
              let inspectorComponent =
                BuildComponentTool.buildInspectorComponent(
                  TestTool.buildAppStateSceneGraphFromEngine(),
                  InspectorTool.buildFakeAllShowComponentConfig(),
                );
              BaseEventTool.triggerComponentEvent(
                inspectorComponent,
                GameObjectRenameTool.triggerRenameChangeEvent(newName),
              );
              BaseEventTool.triggerComponentEvent(
                inspectorComponent,
                GameObjectRenameTool.triggerRenameBlurEvent(newName),
              );

              GameObjectEngineService.unsafeGetGameObjectName(
                SceneEditorService.unsafeGetCurrentSceneTreeNode
                |> StateLogicService.getEditorState,
              )
              |> StateLogicService.getEngineStateToGetData
              |> expect == newName;
            })
          )
        );
      });
      describe("deal with specific case", () => {
        beforeEach(() =>
          MainEditorSceneTool.createDefaultScene(
            sandbox,
            MainEditorSceneTool.setCameraTobeCurrentSceneTreeNode,
          )
        );
        test(
          "test if specific component not exist, should throw error when parse config from json data",
          () =>
          expect(() =>
            BuildComponentTool.buildInspectorComponent(
              TestTool.buildEmptyAppState(),
              InspectorTool.buildFakeErrorAllShowComponentConfig(),
            )
          )
          |> toThrowMessageRe([%re {|/specific\scomponent.+is\serror/mg|}])
        );
        test(
          "test if specific component not exist, should throw error when build component",
          () =>
          expect(() =>
            InspectorTool.buildComponentUIComponent(
              ("SceneTree", 0),
              (TestTool.buildEmptyAppState(), TestTool.getDispatch()),
            )
          )
          |> toThrowMessage("the component: SceneTree not exist")
        );
      });
    });
  });