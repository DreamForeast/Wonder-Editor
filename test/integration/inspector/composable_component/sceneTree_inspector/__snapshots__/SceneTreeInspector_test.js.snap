// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SceneTreeInspector prepare currentSelectSource is SceneTree test gameObject rename test snapshot deal with specific case ket in '', trigger onBlur, the input value should be original name 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="cube"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSelectSource is SceneTree test gameObject rename test snapshot test rename to specific name 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="gameObjectNewName"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSelectSource is SceneTree test show component else else if currentSceneTreeNode is box, should show transform and basicMaterial 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="cube"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSelectSource is SceneTree test show component else if currentSceneTreeNode is camera, should show transform and basicCameraView and perspectiveCameraProjection 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="camera"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="40"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicCameraView
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate basic camera view
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          perspectiveCameraProjection
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate perspective camera view
      </div>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={true}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSelectSource is SceneTree test show component if hasn't currentSceneTreeNode, show nothing 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="cube"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSelectSource is SceneTree test show component test add component workflow click sourceInstance component, add to inspector 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="cube"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          sourceInstance
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate source instance
      </div>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={true}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSelectSource is SceneTree test show component test add component workflow click the add component button, show addableComponent list 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="cube"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
      <div
        onClick={[Function]}
      >
        sourceInstance
      </div>
    </article>
  </article>
</article>
`;
